test/shared/src/main/scala/zio/test/MutableRunnableSpec.scala

```diff
@@ -0,0 +1,144 @@
+package zio.test
+
+import zio.clock.Clock
+import zio.duration._
+import zio.test.environment.TestEnvironment
+import zio.{Chunk, Has, URIO, ZIO, ZLayer}
+
+import scala.util.control.NoStackTrace
+
+/**
+ * Syntax for writing test like
+ * {{{
+ * object MySpec extends MutableRunnableSpec(layer) {
+ *   suite("foo") {
+ *     testM("name") {
+ *     } @@ ignore
+ *
+ *     test("name 2")
+ *   }
+ *   suite("another suite") {
+ *     test("name 3")
+ *   }
+ * }
+ * }}}
+ */
+class MutableRunnableSpec[R <: Has[_]](layer: ZLayer[TestEnvironment, Throwable, R])
+    extends RunnableSpec[TestEnvironment, Any] {
+  self =>
+
+  type ZS = ZSpec[R, Any]
+
+  class InAnotherTestException(`type`: String, label: String)
+      extends Exception(s"${`type`} `${label}` is in another test")
+      with NoStackTrace
+
+  sealed trait SpecBuilder {
+    def toSpec: ZS
+    def label: String
+  }
+
+  case class SuiteBuilder(label: String) extends SpecBuilder {
```

Comment:
`final` or at least `sealed`

Reply:
Fixed

Reply:
I made them just `sealed`. It is fine to have them inside `MutableRunnableSpec` not to pollute `zio.test` package and I can also use type parameters from `MutableRunnableSpec` and do not need to add them to `*Builder` classes. If I make them final, I get this compilation error `The outer reference in this type test cannot be checked at run time.`